### ISSUE TITLE
[question] How to handle test_package while cross compiling a tool package?

### ISSUE THREAD

**Issue by zhanglin-zhou** [conan-io/conan#15548]
### What is your question?

I am trying to build 7zip recipe on Mac, and get test_package error while building x86_64 version with armv8 build profile, my build profile is:
```
[settings]
os=Macos
arch=armv8
```
my host profile is:
```
[settings]
os=Macos
arch=x86_64

compiler=apple-clang
compiler.cppstd=20
compiler.libcxx=libc++
compiler.version=14.0
```
This leads to error while launching test_package:

> ======== Installing packages ========
> ERROR: Missing binary: 7zip/23.01:dcf68e932572755309a5f69f3cee1bede410e907

The reason is: I am creating a x86_64 package per host profile, but in test_package, 7zip is declared as tool:
`        self.tool_requires(self.tested_reference_str)
`
so it uses build profile to find package.

What's the best practice to handle this scenario? If conan cannot support this scenario, is it possible to skip test_package for this cross-compile scenario?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by AbrilRBS** [comment]
Hi! Thanks a lot for your question.

Most Conan commands accept the `--build-require` argument to be used when the package is meant to be a build requirement. For `conan create` and specifically for this case, it has the effect of swapping the profiles in the test_package. Indeed if I try to compile a similar setup, adding said argument fixes the issue and the test runs.

Quick question: Did you modify the 7zip recipe? It currently validates against non-Windows/x86{64} builds, but removing those seem to result in a proper compilation 🤔

**Comment by zhanglin-zhou** [comment]
> Hi! Thanks a lot for your question.
> 
> Most Conan commands accept the `--build-require` argument to be used when the package is meant to be a build requirement. For `conan create` and specifically for this case, it has the effect of swapping the profiles in the test_package. Indeed if I try to compile a similar setup, adding said argument fixes the issue and the test runs.
> 
> Quick question: Did you modify the 7zip recipe? It currently validates against non-Windows/x86{64} builds, but removing those seem to result in a proper compilation 🤔

Yeah, besides removing the validation, I add a little piece of code to make it built on Mac:
```
 def _build_autotools(self):
        # TODO: Enable non-Windows methods in configure
        autotools = Autotools(self)
        if self.settings.os == "Macos":
            with chdir(self, os.path.join(self.source_folder, "CPP", "7zip", "Bundles", "Alone2")):
                autotools.make(args=["-f", "../../cmpl_mac_%s.mak" % ("x64" if self.settings.arch == "x86_64" else "arm64") ], target="all")
        else:
            with chdir(self, os.path.join(self.source_folder, "CPP", "7zip", "Bundles", "LzmaCon")):
                autotools.make(args=["-f", "makefile.gcc"], target="all")
```
It looks like `--build-require` cannot satisfy the requirement in this case, because it creates the package using the configuration and settings of the “build” context, which is "armv8" in this case because my computer is Apple Silicon, so that it will create a binary for armv8, but I want to build a tool which runs on x86_64 computer.

**Comment by memsharded** [comment]
Sorry this was not followed up back then @zhanglin-zhou 

Did you manage to work around the issues?

> It looks like --build-require cannot satisfy the requirement in this case, because it creates the package using the configuration and settings of the “build” context, which is "armv8" in this case because my computer is Apple Silicon, so that it will create a binary for armv8, but I want to build a tool which runs on x86_64 computer.

The ``tool-requires`` themselves cannot be cross-compiled to run in a different platform than the one you are using. They are not "host" packages to cross-compile, so if your current platform is ``armv8`` the binary for 7zip must be a ``armv8`` one. You might use some of the mechanisms to define binary compatibility between ``armv8`` binaries and ``x86_64`` binaries, so Conan might be able to fallback to a ``x86_64`` binary when you are in a ``armv8`` machine, but you cannot directly tool-require such package and have it cross-built to a different platform than the current "build" one.
You might be able to do it in 2 steps, in which first you cross-build your package independently to the ``x86_64`` architecture, and then use it as a pre-compiled binary with the above mentioned binary ``compatibility`` mechanisms.

In practice it might mean to use ``self.requires(self.tested_reference_str)`` instead of the ``self.tool_requires(self.tested_reference_str)``, and use cross-build profiles as usual.

Not sure if it is worth the effort, why not just building and using an ``armv8`` executable binary for your ``tool-require``?

**Comment by memsharded** [comment]
Closing this ticket as staled, please create new tickets for any further question or issue, thanks for the feedback!
